From 5dc2988f4b90a4464597dfa65fb203a6be8b17d5 Mon Sep 17 00:00:00 2001
From: "Lori A. Burns" <lori.burns7@gmail.com>
Date: Mon, 3 Aug 2026 19:27:49 -0400
Subject: [PATCH 1/1] treat pylibxc as independent project2

fit better into upstream Py/CMake system

interp only

upgrade from LOCATION

fine tuning
---
 CMakeLists.txt         | 23 +-------------
 pylibxc/CMakeLists.txt | 70 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 71 insertions(+), 22 deletions(-)
 create mode 100644 pylibxc/CMakeLists.txt

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8195880f7..04c6dfd0c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -504,28 +504,7 @@ if(ENABLE_FORTRAN)
 endif()
 
 if(ENABLE_PYTHON)
-  set(SOURCE_PYTHON_API
-    pylibxc/__init__.py
-    pylibxc/array_backend.py
-    pylibxc/core.py
-    pylibxc/dlpack.py
-    pylibxc/flags.py
-    pylibxc/functional.py
-    pylibxc/structs.py
-    pylibxc/testing.py
-    pylibxc/util.py
-    pylibxc/version.py
-    pylibxc/example_densities.py
-  )
-  install(
-    FILES ${SOURCE_PYTHON_API}
-    DESTINATION ${CMAKE_INSTALL_LIBDIR}/pylibxc
-  )
-  if(UNIX)
-    install(CODE "execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink \
-      ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}/libxc.so.${${PROJECT_NAME}_SOMAJOR} \
-      ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}/pylibxc/libxc.so)")
-  endif()
+  add_subdirectory(pylibxc)
 endif()
 
 export(EXPORT c_interface
diff --git a/pylibxc/CMakeLists.txt b/pylibxc/CMakeLists.txt
new file mode 100644
index 000000000..1bec25009
--- /dev/null
+++ b/pylibxc/CMakeLists.txt
@@ -0,0 +1,70 @@
+cmake_minimum_required(VERSION 3.21)
+cmake_policy(SET CMP0079 NEW)
+
+project(pylibxc LANGUAGES NONE)
+list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/../cmake)
+
+################################### Options ####################################
+include(psi4OptionsTools)
+option_with_default(PYLIBXC_INSTALL_PYMODDIR "Location within CMAKE_INSTALL_PREFIX to which the Python module is installed. Empty string queries Python interpreter. Don't start with '/'" prefix)
+
+################################# Main Project #################################
+
+#  <<<  Build  >>>
+
+if (PYLIBXC_INSTALL_PYMODDIR STREQUAL "prefix")
+
+    if (NOT TARGET Python3::Interpreter)
+        find_package(Python3 COMPONENTS Interpreter REQUIRED)
+    endif()
+
+    # Note that this block is *Linux-style* install to `CMAKE_INSTALL_PREFIX` not *Python-style* install to `Python3_EXECUTABLE`'s site-packages.
+
+    execute_process(
+      COMMAND ${Python3_EXECUTABLE} -c
+        "import os, sys, sysconfig as s; spdir = s.get_path('platlib'); print(spdir.replace(os.path.commonpath([sys.prefix, spdir]), '').lstrip(os.path.sep));"
+      OUTPUT_VARIABLE PYLIBXC_INSTALL_PYMODDIR
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+      )
+endif()
+message(STATUS "Showing option PYLIBXC_INSTALL_PYMODDIR: ${PYLIBXC_INSTALL_PYMODDIR}")
+
+if(TARGET xc)
+    set(_xc_target xc)
+else()
+    find_package(Libxc REQUIRED COMPONENTS C shared)
+    set(_xc_target Libxc::xc)
+endif()
+
+#  <<<  Install  >>>
+
+install(
+  FILES
+    __init__.py
+    array_backend.py
+    core.py
+    dlpack.py
+    flags.py
+    functional.py
+    structs.py
+    testing.py
+    util.py
+    version.py
+    example_densities.py
+  COMPONENT Libxc_Python
+  DESTINATION ${PYLIBXC_INSTALL_PYMODDIR}/pylibxc
+  )
+
+if (UNIX)
+  install(CODE "
+    file(MAKE_DIRECTORY
+      \"\$ENV{DESTDIR}\${CMAKE_INSTALL_PREFIX}/${PYLIBXC_INSTALL_PYMODDIR}/pylibxc\"
+    )
+    execute_process(
+      COMMAND \"${CMAKE_COMMAND}\" -E create_symlink
+              \"\$ENV{DESTDIR}\${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}/$<TARGET_FILE_NAME:${_xc_target}>\"
+              \"\$ENV{DESTDIR}\${CMAKE_INSTALL_PREFIX}/${PYLIBXC_INSTALL_PYMODDIR}/pylibxc/libxc.so\"
+      COMMAND_ERROR_IS_FATAL ANY
+    )
+  ")
+endif()
-- 
2.43.0

